src/sidebar/test/permissions-test.js

```diff
@@ -0,0 +1,90 @@
+'use strict';
+
+var Permissions = require('../permissions');
+
+var userid = 'acct:flash@gord.on';
+
+describe('permissions', function () {
+  var fakeLocalStorage;
+  var permissions;
+
+  beforeEach(function () {
+    fakeLocalStorage = {
+      getItem: sinon.stub().returns(null),
+      setItem: sinon.stub(),
+    };
+    permissions = new Permissions(fakeLocalStorage);
+  });
+
+  describe('#private', function () {
+    it('only allows the user to read the annotation', function () {
+      assert.deepEqual(permissions.private(userid), {
+        read: [userid],
+        update: [userid],
+        delete: [userid],
+      });
+    });
+  });
+
+  describe('#shared', function () {
+    it('allows the group to read the annotation', function () {
+      assert.deepEqual(permissions.shared(userid, 'gid'), {
+        read: ['group:gid'],
+        update: [userid],
+        delete: [userid],
+      });
+    });
+  });
+
+  describe('#default', function () {
+    it('returns shared permissions by default', function () {
+      assert.deepEqual(permissions.default(userid, 'gid'),
+                       permissions.shared(userid, 'gid'));
+    });
+
+    it('returns private permissions if the saved level is "private"', function () {
+      fakeLocalStorage.getItem.returns('private');
+      assert.deepEqual(permissions.default(userid, 'gid'),
+                       permissions.private(userid));
+    });
+
+    it('returns shared permissions if the saved level is "shared"', function () {
+      fakeLocalStorage.getItem.returns('shared');
+      assert.deepEqual(permissions.default(userid, 'gid'),
+                       permissions.shared(userid, 'gid'));
+    });
+  });
+
+  describe('#setDefault', function () {
+    it('saves the default permissions', function () {
+      permissions.setDefault('private');
+      assert.calledWith(fakeLocalStorage.setItem, 'hypothesis.privacy', 'private');
+    });
+  });
+
+  describe('#isShared', function () {
+    it('returns true if a group can read the annotation', function () {
+      var perms = permissions.shared(userid, 'gid');
+      assert.isTrue(permissions.isShared(perms));
```

Comment:
This test and a number of ones above and below, such as those for `#default` and `#permits`, are interesting. I think they're good tests. But this one for example is not really a unit test for `isShared()`. It's a test for both `shared()` and `isShared()` together that would fail if either of them was wrong. This is actually a test of the **contract** that `permissions` provides, beyond its **interface** (what methods it has and their arguments, return values and exceptions) it also provides a bunch of cross-method **invariants** such as **If you create a `Permissions` object by calling `shared()` then passing that same object to `isShared()` will return true**. Several of the tests here are testing these sorts of cross-method invariants.

I think these are actually contract tests rather than unit tests, and I think contract tests are a great thing to have in addition to unit tests, but I don't think we have a good naming convention for contract tests either in our Python or our JavaScript code. I wonder if it should be something like:

```javascript
describe('src.sidebar.permissions', function() {

  ...

  describe("permissions's contract", function() {
    specify('Permissions from shared() return true if passed to isShared()', function() {
```